From 6389031d394f1d0a556cd158710c8eeb76dd7d3c Mon Sep 17 00:00:00 2001
From: Francois Choleau <francois.choleau@st.com>
Date: Wed, 3 Apr 2024 03:20:37 -0400
Subject: [PATCH 0808/1695] plat-stm32mp2: stm32mp25_pwr: manage CID filtering
 when accessing PWR_CR12

There is no inheritance between PWR_CR12 and GPU RIFSC resources.
Disabling/re-enabling CID filtering of GPU RIFSC
before/after each access of PWR_CR12 (read or write) fixes
this lack.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2c0aa8f3db29f23c0ec7cefc02e649f9f0fa4fe3 plat-stm32mp2: stm32mp25_pwr: manage CID filtering when accessing PWR_CR12]
Change-Id: Iaa0de73e7849fdd4f3f656beb54c583cfd0fe99a
Signed-off-by: Francois Choleau <francois.choleau@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/371168
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/375006
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 .../drivers/stm32mp25_pwr_regul.c             | 34 +++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_regul.c b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_regul.c
index bf39309f2..a6ef28c56 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_regul.c
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_regul.c
@@ -7,6 +7,7 @@
 #include <compiler.h>
 #include <drivers/regulator.h>
 #include <drivers/stm32mp25_pwr.h>
+#include <drivers/stm32_rif.h>
 #include <io.h>
 #include <kernel/delay.h>
 #include <kernel/dt.h>
@@ -98,16 +99,30 @@ struct pwr_regu {
 	 */
 	bool is_an_iod;
 	bool keep_monitor_on;
+	/*
+	 * rifsc_filtering_id is used to disable filtering when
+	 * accessing to the register
+	 */
+	uint8_t rifsc_filtering_id;
 };
 
 static TEE_Result pwr_enable_reg(struct pwr_regu *pwr_regu)
 {
 	uintptr_t reg = stm32_pwr_base() + pwr_regu->enable_reg;
 	uint64_t to = 0;
+	bool cid_enabled = false;
 
 	if (!pwr_regu->enable_mask)
 		return TEE_SUCCESS;
 
+	if (pwr_regu->rifsc_filtering_id)
+		cid_enabled =
+			stm32_rifsc_cid_is_enabled
+				(pwr_regu->rifsc_filtering_id);
+
+	if (cid_enabled)
+		stm32_rifsc_cid_disable(pwr_regu->rifsc_filtering_id);
+
 	io_setbits32(reg, pwr_regu->enable_mask);
 
 	/* Wait for vddgpu to enable as stated in the reference manual */
@@ -121,6 +136,8 @@ static TEE_Result pwr_enable_reg(struct pwr_regu *pwr_regu)
 
 	if (!(io_read32(reg) & pwr_regu->ready_mask)) {
 		io_clrbits32(reg, pwr_regu->enable_mask);
+		if (cid_enabled)
+			stm32_rifsc_cid_enable(pwr_regu->rifsc_filtering_id);
 		return TEE_ERROR_GENERIC;
 	}
 
@@ -130,18 +147,34 @@ static TEE_Result pwr_enable_reg(struct pwr_regu *pwr_regu)
 	if (!pwr_regu->keep_monitor_on)
 		io_clrbits32(reg, pwr_regu->enable_mask);
 
+	if (cid_enabled)
+		stm32_rifsc_cid_enable(pwr_regu->rifsc_filtering_id);
+
 	return TEE_SUCCESS;
 }
 
 static void pwr_disable_reg(struct pwr_regu *pwr_regu)
 {
 	uintptr_t reg = stm32_pwr_base() + pwr_regu->enable_reg;
+	bool cid_enabled = false;
 
 	if (pwr_regu->enable_mask) {
+		if (pwr_regu->rifsc_filtering_id)
+			cid_enabled =
+				stm32_rifsc_cid_is_enabled
+					(pwr_regu->rifsc_filtering_id);
+
+		if (cid_enabled)
+			stm32_rifsc_cid_disable(pwr_regu->rifsc_filtering_id);
+
 		/* Make sure the previous operations are visible */
 		dsb();
 		io_clrbits32(reg, pwr_regu->enable_mask | pwr_regu->valid_mask);
 	}
+
+	if (cid_enabled)
+		stm32_rifsc_cid_enable(pwr_regu->rifsc_filtering_id);
+
 }
 
 static TEE_Result pwr_set_state(struct regulator *regulator, bool enable)
@@ -461,6 +494,7 @@ static struct pwr_regu pwr_regulators[PWR_REGU_COUNT] = {
 		.ready_mask = PWR_CR12_VDDGPURDY,
 		.valid_mask = PWR_CR12_GPUSV,
 		.keep_monitor_on = true,
+		.rifsc_filtering_id = STM32MP25_RIFSC_GPU_ID,
 	 },
 };
 
-- 
2.47.2

